esp/services/ws_dfu/ws_dfuService.cpp

```diff
@@ -6054,50 +6084,35 @@ void CWsDfuEx::getFileAccess(IEspContext &context, IUserDescriptor *udesc, SecAc
         }
         case CFileAccessRole_External:
         {
-            IArrayOf<IEspDFUPartLocations> dfuPartLocations;
-            IArrayOf<IEspDFUPartCopies> dfuPartCopies;
-            resp.setNumParts(getFilePartsInfo(context, df, req.getCluster(), dfuPartLocations, dfuPartCopies));
-            resp.setFilePartLocations(dfuPartLocations);
-            resp.setFileParts(dfuPartCopies);
+            getFilePartsInfo(context, fileDesc, df->numParts(), false, accessInfo);
             if (req.getReturnJsonTypeInfo() || req.getReturnJsonTypeInfo())
             {
                 MemoryBuffer binLayout;
                 StringBuffer jsonLayout;
                 if (!getRecordFormatFromRtlType(binLayout, jsonLayout, df->queryAttributes(), req.getReturnJsonTypeInfo(), req.getReturnJsonTypeInfo()))
                     getRecordFormatFromECL(binLayout, jsonLayout, df->queryAttributes(), req.getReturnJsonTypeInfo(), req.getReturnJsonTypeInfo());
                 if (req.getReturnJsonTypeInfo() && jsonLayout.length())
-                    resp.setRecordTypeInfoJson(jsonLayout.str());
+                    accessInfo.setRecordTypeInfoJson(jsonLayout.str());
                 if (req.getReturnBinTypeInfo() && binLayout.length())
-                    resp.setRecordTypeInfoBin(binLayout);
+                    accessInfo.setRecordTypeInfoBin(binLayout);
             }
             break;
         }
         default:
             throwUnexpected();
     }
 
-    // setup "expiryTime"
-    unsigned expirySecs = req.getExpirySeconds();
-    if (expirySecs > maxFileAccessExpirySeconds)
-        expirySecs = maxFileAccessExpirySeconds;
-    time_t now;
-    time(&now);
-    CDateTime expiryDt;
-    expiryDt.set(now + expirySecs);
-    StringBuffer expiryTimeStr;
-    expiryDt.getString(expiryTimeStr);
+    StringBuffer metaInfo, expiryTime;
+    getFileMeta(metaInfo, expiryTime, fileName, fileDesc, udesc, req.getJobId(), req);
+    accessInfo.setMetaInfoBlob(metaInfo);
+    accessInfo.setExpiryTime(expiryTime);
 
     StringBuffer keyPairName;
     unsigned port;
     bool secure;
     getFileDafilesrvConfiguration(keyPairName, port, secure, *df);
```

Comment:
@wangkx - something wrong here.. this used to be before the call to getFilemeta().

getFileDafilesrvConfiguration gets the keyPairName associated with the file (the cluster the file is on).

That must then be used to sign.
The req.getCluster() should only be used (if present) to lookup the logical file.

Reply:
Ok, I will move line 6110-6113 before line 6105.

Reply:
@jakesmith The getFileMeta() is also called when DFUFileCreate().  Should I call the getFileDafilesrvConfiguration() to get the keyPairName when DFUFileCreate()?

Reply:
> @jakesmith The getFileMeta() is also called when DFUFileCreate(). Should I call the getFileDafilesrvConfiguration() to get the keyPairName when DFUFileCreate()?

The key difference on create vs read, is that create you have to say where you want it created (you have to provide the cluster group).
You still need to get the dafilersv details, but you already have the cluster.
You could reuse getFileDafilesrvConfiguration() if you call it after you create the IDistributeFile possibly, but really it should be getting the cluster from the file it just created.

Reply:
I would split getFileDafilesrvConfiguration(), so that you can call it with a named cluster

e.g. (NB: haven't checked it compiles!)
```
void CWsDfuEx::getFileDafilesrvConfiguration(StringBuffer &keyPairName, unsigned &retPort, bool &retSecure, const char *cluster)
{
    retPort = DEFAULT_ROWSERVICE_PORT;
    retSecure = false;
    const char *_keyPairName = env->getClusterKeyPairName(cluster);
    Owned<IConstDaFileSrvInfo> daFileSrvInfo = env->getDaFileSrvGroupInfo(cluster);
    unsigned port = DEFAULT_ROWSERVICE_PORT;
    bool secure = false;
    if (daFileSrvInfo)
    {
        port = daFileSrvInfo->getPort();
        secure = daFileSrvInfo->getSecure();
    }
}

void CWsDfuEx::getFileDafilesrvConfiguration(StringBuffer &keyPairName, unsigned &retPort, bool &retSecure, IDistributedFile &file)
{
    retPort = DEFAULT_ROWSERVICE_PORT;
    retSecure = false;
    unsigned numClusters = file.numClusters();
    for (unsigned c=0; c<numClusters; c++)
    {
        unsigned port = DEFAULT_ROWSERVICE_PORT;
        bool secure = false;

        StringBuffer clusterName;
        const char *cluster = file.getClusterName(c, clusterName.clear()).str();
        
        StringBuffer _keyPairName;
        getFileDafilesrvConfiguration(_keyPairName, port, secure);     
        if (0 == c)
        {
            keyPairName.set(_keyPairName);
            retPort = port;
            retSecure = secure;
        }
        else
        {
            if (!strsame(keyPairName, _keyPairName))
                throwStringExceptionV(0, "Configuration issue - file '%s' is on multiple clusters, keys for file access must match", file.queryLogicalName());
            if (retPort != port)
                throwStringExceptionV(0, "Configuration issue - file '%s' is on multiple clusters, dafilesrv's port for file access must match", file.queryLogicalName());
            if (retSecure != secure)
                throwStringExceptionV(0, "Configuration issue - file '%s' is on multiple clusters, dafilesrv's security setting for file access must match", file.queryLogicalName());
        }
    }
}
```

Reply:
@wangkx - back to you.